cpp/src/arrow/array/concatenate.cc

```diff
@@ -133,26 +159,30 @@ Status ConcatenateOffsets(const BufferVector& buffers, MemoryPool* pool,
   for (size_t i = 0; i < buffers.size(); ++i) {
     // the first offset from buffers[i] will be adjusted to values_length
     // (the cumulative length of values spanned by offsets in previous buffers)
-    RETURN_NOT_OK(PutOffsets<Offset>(*buffers[i], values_length,
-                                     out_data + elements_length, &(*values_ranges)[i]));
+    ARROW_ASSIGN_OR_RAISE(auto outcome, PutOffsets<Offset>(*buffers[i], values_length,
+                                                           out_data + elements_length,
+                                                           &(*values_ranges)[i]));
+    if (ARROW_PREDICT_FALSE(outcome != OffsetBufferOpOutcome::kOk)) {
```

Comment:
I know the reason why `RETURN_IF_NOT_OK_OUTCOME` not used but actually it's a bitweird 🤔